synapse/storage/databases/main/events_worker.py

```diff
@@ -608,13 +615,36 @@ def _get_events_from_cache(
         event_map = {}
 
         for event_id in events:
+            # First check if its in the event cache
             ret = self._get_event_cache.get(
                 (event_id,), None, update_metrics=update_metrics
             )
-            if not ret:
-                continue
+            if ret:
+                event_map[event_id] = ret
+
+            # Otherwise check if we still have the event in memory.
+            event = self._event_ref.get(event_id)
+            if event:
+                # Reconstruct an event cache entry
+
+                redacted_event = None
+                if event.internal_metadata.redacted_by is not None:
+                    # The event has been redacted, so we generate a redacted
+                    # version.
+                    redacted_event = prune_event(event)
+                    redacted_event.unsigned[
+                        "redacted_by"
+                    ] = event.internal_metadata.redacted_by
+
+                cache_entry = _EventCacheEntry(
+                    event=event,
+                    redacted_event=redacted_event,
+                )
+                event_map[event_id] = cache_entry
 
-            event_map[event_id] = ret
+                # We add the entry back into the cache as we want to keep
+                # recently queried events in the cache.
+                self._get_event_cache.set((event_id,), cache_entry)
```

Comment:
From https://github.com/matrix-org/synapse/pull/10119#discussion_r681146647:

> could we refactor the similar code in `_get_events_from_db`/`_maybe_redact_event_row` so that we can reuse some common code rather than have two similar but slightly different impls?
> 
> For example: have `_maybe_redact_event_row` just return a boolean rather than the redaction; have `_get_events_from_db` poke `internal_metadata.redacted_by` based on that boolean; then move these lines here out to a `build_cache_entry` which can be called by `_get_events_from_db`.